src/floating.cpp

```diff
@@ -14,30 +14,35 @@
 
 using namespace std;
 
-void floating_init() {
+Direction::Direction(char ch)
+{
+    std::map<char, V> mapping = {
+        {'u', Up}, {'r', Right}, {'d', Down}, {'l', Left},
+    };
+    auto it = mapping.find(ch);
+    if (it != mapping.end()) {
+        value = it->second;
+        valid = true;
```

Comment:
Is indicating a parsing failure within the ctor the only purpose of the `valid` member?
There is no default ctor being used anywhere, right?
If so, I'd suggest throwing an exception here instead, and get rid of the valid flag.

(If there is any need for having "undefined, invalid" instances of `Direction`, I would recommend using `std::optional`)